File: components/cell-mesh/src/cell_channel.c
// cell_channel.c — payment-channel state machine with linearity
// enforcement. Pure C, no IDF dependency.

#include "cell_channel.h"
#include "cell_wire.h"

#include <string.h>

// ── Wire encoding offsets ──────────────────────────────────────────────

// channel_open (61 bytes)
#define O_OFF_CHANNEL_ID         0u
#define O_OFF_PEER_PUBKEY        16u
#define O_OFF_LOCKTIME_MS        49u
#define O_OFF_CAPACITY           57u

// channel_commitment (68 bytes)
#define C_OFF_CHANNEL_ID         0u
#define C_OFF_SEQ                16u
#define C_OFF_DEVICE_SHARE       20u
#define C_OFF_USER_SHARE         24u
#define C_OFF_EXPIRY_MS          28u
#define C_OFF_CERT_HASH          36u  // 32 bytes — SHA-256(cap cert payload)

// channel_close (24 bytes)
#define K_OFF_CHANNEL_ID         0u
#define K_OFF_FINAL_SEQ          16u
#define K_OFF_FINAL_DEVICE_SHARE 20u

// ── Encoders ──────────────────────────────────────────────────────────

int cm_channel_open_encode(const cm_channel_open_t *in,
                           uint8_t out[CM_CHANNEL_OPEN_PAYLOAD_BYTES]) {
    if (!in || !out) return -1;
    memcpy(out + O_OFF_CHANNEL_ID,  in->channel_id,  16);
    memcpy(out + O_OFF_PEER_PUBKEY, in->peer_pubkey, 33);
    cm_write_u64(out + O_OFF_LOCKTIME_MS, in->initial_locktime_ms);
    cm_write_u32(out + O_OFF_CAPACITY,    in->total_capacity);
    return 0;
}

int cm_channel_open_decode(const uint8_t in[CM_CHANNEL_OPEN_PAYLOAD_BYTES],
                           cm_channel_open_t *out) {
    if (!in || !out) return -1;
    memcpy(out->channel_id,  in + O_OFF_CHANNEL_ID,  16);
    memcpy(out->peer_pubkey, in + O_OFF_PEER_PUBKEY, 33);
    out->initial_locktime_ms = cm_read_u64(in + O_OFF_LOCKTIME_MS);
    out->total_capacity      = cm_read_u32(in + O_OFF_CAPACITY);
    return 0;
}

int cm_channel_commitment_encode(const cm_channel_commitment_t *in,
                                 uint8_t out[CM_CHANNEL_COMMITMENT_PAYLOAD_BYTES]) {
    if (!in || !out) return -1;
    memcpy(out + C_OFF_CHANNEL_ID, in->channel_id, 16);
    cm_write_u32(out + C_OFF_SEQ,          in->seq);
    cm_write_u32(out + C_OFF_DEVICE_SHARE, in->device_share);
    cm_write_u32(out + C_OFF_USER_SHARE,   in->user_share);
    cm_write_u64(out + C_OFF_EXPIRY_MS,    in->expiry_ms);
    memcpy(out + C_OFF_CERT_HASH, in->cert_hash, 32);
    return 0;
}

int cm_channel_commitment_decode(const uint8_t in[CM_CHANNEL_COMMITMENT_PAYLOAD_BYTES],
                                 cm_channel_commitment_t *out) {
    if (!in || !out) return -1;
    memcpy(out->channel_id, in + C_OFF_CHANNEL_ID, 16);
    out->seq          = cm_read_u32(in + C_OFF_SEQ);
    out->device_share = cm_read_u32(in + C_OFF_DEVICE_SHARE);
    out->user_share   = cm_read_u32(in + C_OFF_USER_SHARE);
    out->expiry_ms    = cm_read_u64(in + C_OFF_EXPIRY_MS);
    memcpy(out->cert_hash, in + C_OFF_CERT_HASH, 32);
    return 0;
}

int cm_channel_close_encode(const cm_channel_close_t *in,
                            uint8_t out[CM_CHANNEL_CLOSE_PAYLOAD_BYTES]) {
    if (!in || !out) return -1;
    memcpy(out + K_OFF_CHANNEL_ID, in->channel_id, 16);
    cm_write_u32(out + K_OFF_FINAL_SEQ,           in->final_seq);
    cm_write_u32(out + K_OFF_FINAL_DEVICE_SHARE,  in->final_device_share);
    return 0;
}

int cm_channel_close_decode(const uint8_t in[CM_CHANNEL_CLOSE_PAYLOAD_BYTES],
                            cm_channel_close_t *out) {
    if (!in || !out) return -1;
    memcpy(out->channel_id, in + K_OFF_CHANNEL_ID, 16);
    out->final_seq          = cm_read_u32(in + K_OFF_FINAL_SEQ);
    out->final_device_share = cm_read_u32(in + K_OFF_FINAL_DEVICE_SHARE);
    return 0;
}

// ── State machine ─────────────────────────────────────────────────────

void cm_channel_init(cm_channel_t *c) {
    if (!c) return;
    memset(c, 0, sizeof(*c));
    c->state = CM_CHAN_CLOSED;
}

cm_channel_rc_t cm_channel_apply_open(cm_channel_t *c,
                                       const cm_channel_open_t *op) {
    if (!c || !op) return CM_CHAN_ERR_BAD_STATE;
    if (c->state != CM_CHAN_CLOSED) return CM_CHAN_ERR_BAD_STATE;

    memcpy(c->channel_id,  op->channel_id,  16);
    memcpy(c->peer_pubkey, op->peer_pubkey, 33);
    c->total_capacity   = op->total_capacity;
    c->open_locktime_ms = op->initial_locktime_ms;

    // Reset commitment fields — no commitment yet.
    c->current_seq         = 0;
    c->device_share        = 0;
    c->user_share          = 0;
    c->expiry_ms           = 0;

    c->state = CM_CHAN_OPEN;
    return CM_CHAN_OK;
}

cm_channel_rc_t cm_channel_apply_commitment(cm_channel_t *c,
                                             const cm_channel_commitment_t *cm,
                                             uint64_t now_ms) {
    if (!c || !cm) return CM_CHAN_ERR_BAD_STATE;
    if (c->state != CM_CHAN_OPEN && c->state != CM_CHAN_ACTIVE) {
        c->commitments_rejected++;
        return CM_CHAN_ERR_BAD_STATE;
    }
    if (memcmp(c->channel_id, cm->channel_id, 16) != 0) {
        c->commitments_rejected++;
        return CM_CHAN_ERR_BAD_ID;
    }
    // Linearity: seq must strictly increase.
    if (cm->seq <= c->current_seq) {
        c->commitments_rejected++;
        return CM_CHAN_ERR_STALE_SEQ;
    }
    // Linearity: device_share must monotonically non-decrease.
    if (cm->device_share < c->device_share) {
        c->commitments_rejected++;
        return CM_CHAN_ERR_NON_MONO;
    }
    // Sanity: device + user must not exceed total_capacity.
    if ((uint64_t)cm->device_share + (uint64_t)cm->user_share > (uint64_t)c->total_capacity) {
        c->commitments_rejected++;
        return CM_CHAN_ERR_OVERFLOW;
    }
    // Expiry must be in the future.
    if (cm->expiry_ms <= now_ms) {
        c->commitments_rejected++;
        return CM_CHAN_ERR_EXPIRED;
    }

    c->current_seq  = cm->seq;
    c->device_share = cm->device_share;
    c->user_share   = cm->user_share;
    c->expiry_ms    = cm->expiry_ms;
    c->commitments_received++;
    c->state = CM_CHAN_ACTIVE;
    return CM_CHAN_OK;
}

cm_channel_rc_t cm_channel_apply_close(cm_channel_t *c,
                                        const cm_channel_close_t *cl) {
    if (!c || !cl) return CM_CHAN_ERR_BAD_STATE;
    if (c->state != CM_CHAN_ACTIVE && c->state != CM_CHAN_EXPIRED) {
        return CM_CHAN_ERR_BAD_STATE;
    }
    if (memcmp(c->channel_id, cl->channel_id, 16) != 0) {
        return CM_CHAN_ERR_BAD_ID;
    }
    // Close must reference the current commitment exactly.
    if (cl->final_seq != c->current_seq
        || cl->final_device_share != c->device_share) {
        return CM_CHAN_ERR_SEQ_MATCH;
    }
    c->state = CM_CHAN_CLOSED;
    return CM_CHAN_OK;
}

void cm_channel_tick_expiry(cm_channel_t *c, uint64_t now_ms) {
    if (!c) return;
    if (c->state == CM_CHAN_ACTIVE && now_ms > c->expiry_ms) {
        c->state = CM_CHAN_EXPIRED;
    }
}

// ── BSV on-chain binding helpers ─────────────────────────────────────

bool cm_channel_validate_utxo_ref(const uint8_t channel_id[16],
                                   const uint8_t txid_display[32]) {
    if (!channel_id || !txid_display) return false;
    // channel_id is derived as the first 16 bytes of the funding txid in
    // display (hex) order — the same order ARC /v1/tx returns it.
    return memcmp(channel_id, txid_display, 16) == 0;
}

File: components/cell-mesh-zig/src/cell_channel.zig
const std = @import("std");
const wire = @import("cell_wire.zig");

const state_closed: c_int = 0;
const state_open: c_int = 1;
const state_active: c_int = 2;
const state_expired: c_int = 3;

const ok: c_int = 0;
const err_bad_state: c_int = -1;
const err_bad_id: c_int = -2;
const err_stale_seq: c_int = -3;
const err_non_mono: c_int = -4;
const err_overflow: c_int = -5;
const err_expired: c_int = -6;
const err_seq_match: c_int = -7;

pub const Channel = extern struct {
    state: c_int,
    channel_id: [16]u8,
    peer_pubkey: [33]u8,
    total_capacity: u32,
    open_locktime_ms: u64,
    current_seq: u32,
    device_share: u32,
    user_share: u32,
    expiry_ms: u64,
    commitments_received: u32,
    commitments_rejected: u32,
};

pub const ChannelOpen = extern struct {
    channel_id: [16]u8,
    peer_pubkey: [33]u8,
    initial_locktime_ms: u64,
    total_capacity: u32,
};

pub const ChannelCommitment = extern struct {
    channel_id: [16]u8,
    seq: u32,
    device_share: u32,
    user_share: u32,
    expiry_ms: u64,
    cert_hash: [32]u8,
};

pub const ChannelClose = extern struct {
    channel_id: [16]u8,
    final_seq: u32,
    final_device_share: u32,
};

pub export fn cm_channel_open_encode(maybe_in: ?*const ChannelOpen, maybe_out: ?[*]u8) callconv(.c) c_int {
    const in = maybe_in orelse return -1;
    const out = maybe_out orelse return -1;
    @memcpy(out[0..16], in.channel_id[0..]);
    @memcpy(out[16..49], in.peer_pubkey[0..]);
    wire.writeU64(out[49..][0..8], in.initial_locktime_ms);
    wire.writeU32(out[57..][0..4], in.total_capacity);
    return 0;
}

pub export fn cm_channel_open_decode(maybe_in: ?[*]const u8, maybe_out: ?*ChannelOpen) callconv(.c) c_int {
    const in = maybe_in orelse return -1;
    const out = maybe_out orelse return -1;
    @memcpy(out.channel_id[0..], in[0..16]);
    @memcpy(out.peer_pubkey[0..], in[16..49]);
    out.initial_locktime_ms = wire.readU64(in[49..][0..8]);
    out.total_capacity = wire.readU32(in[57..][0..4]);
    return 0;
}

pub export fn cm_channel_commitment_encode(
    maybe_in: ?*const ChannelCommitment,
    maybe_out: ?[*]u8,
) callconv(.c) c_int {
    const in = maybe_in orelse return -1;
    const out = maybe_out orelse return -1;
    @memcpy(out[0..16], in.channel_id[0..]);
    wire.writeU32(out[16..][0..4], in.seq);
    wire.writeU32(out[20..][0..4], in.device_share);
    wire.writeU32(out[24..][0..4], in.user_share);
    wire.writeU64(out[28..][0..8], in.expiry_ms);
    @memcpy(out[36..68], in.cert_hash[0..]);
    return 0;
}

pub export fn cm_channel_commitment_decode(
    maybe_in: ?[*]const u8,
    maybe_out: ?*ChannelCommitment,
) callconv(.c) c_int {
    const in = maybe_in orelse return -1;
    const out = maybe_out orelse return -1;
    @memcpy(out.channel_id[0..], in[0..16]);
    out.seq = wire.readU32(in[16..][0..4]);
    out.device_share = wire.readU32(in[20..][0..4]);
    out.user_share = wire.readU32(in[24..][0..4]);
    out.expiry_ms = wire.readU64(in[28..][0..8]);
    @memcpy(out.cert_hash[0..], in[36..68]);
    return 0;
}

pub export fn cm_channel_close_encode(maybe_in: ?*const ChannelClose, maybe_out: ?[*]u8) callconv(.c) c_int {
    const in = maybe_in orelse return -1;
    const out = maybe_out orelse return -1;
    @memcpy(out[0..16], in.channel_id[0..]);
    wire.writeU32(out[16..][0..4], in.final_seq);
    wire.writeU32(out[20..][0..4], in.final_device_share);
    return 0;
}

pub export fn cm_channel_close_decode(maybe_in: ?[*]const u8, maybe_out: ?*ChannelClose) callconv(.c) c_int {
    const in = maybe_in orelse return -1;
    const out = maybe_out orelse return -1;
    @memcpy(out.channel_id[0..], in[0..16]);
    out.final_seq = wire.readU32(in[16..][0..4]);
    out.final_device_share = wire.readU32(in[20..][0..4]);
    return 0;
}

pub export fn cm_channel_init(maybe_c: ?*Channel) callconv(.c) void {
    const c = maybe_c orelse return;
    c.* = std.mem.zeroes(Channel);
    c.state = state_closed;
}

pub export fn cm_channel_apply_open(maybe_c: ?*Channel, maybe_op: ?*const ChannelOpen) callconv(.c) c_int {
    const c = maybe_c orelse return err_bad_state;
    const op = maybe_op orelse return err_bad_state;
    if (c.state != state_closed) return err_bad_state;

    c.channel_id = op.channel_id;
    c.peer_pubkey = op.peer_pubkey;
    c.total_capacity = op.total_capacity;
    c.open_locktime_ms = op.initial_locktime_ms;
    c.current_seq = 0;
    c.device_share = 0;
    c.user_share = 0;
    c.expiry_ms = 0;
    c.state = state_open;
    return ok;
}

pub export fn cm_channel_apply_commitment(
    maybe_c: ?*Channel,
    maybe_cm: ?*const ChannelCommitment,
    now_ms: u64,
) callconv(.c) c_int {
    const c = maybe_c orelse return err_bad_state;
    const cm = maybe_cm orelse return err_bad_state;
    if (c.state != state_open and c.state != state_active) {
        c.commitments_rejected +%= 1;
        return err_bad_state;
    }
    if (!std.mem.eql(u8, c.channel_id[0..], cm.channel_id[0..])) {
        c.commitments_rejected +%= 1;
        return err_bad_id;
    }
    if (cm.seq <= c.current_seq) {
        c.commitments_rejected +%= 1;
        return err_stale_seq;
    }
    if (cm.device_share < c.device_share) {
        c.commitments_rejected +%= 1;
        return err_non_mono;
    }
    if (@as(u64, cm.device_share) + @as(u64, cm.user_share) > @as(u64, c.total_capacity)) {
        c.commitments_rejected +%= 1;
        return err_overflow;
    }
    if (cm.expiry_ms <= now_ms) {
        c.commitments_rejected +%= 1;
        return err_expired;
    }

    c.current_seq = cm.seq;
    c.device_share = cm.device_share;
    c.user_share = cm.user_share;
    c.expiry_ms = cm.expiry_ms;
    c.commitments_received +%= 1;
    c.state = state_active;
    return ok;
}

pub export fn cm_channel_apply_close(maybe_c: ?*Channel, maybe_cl: ?*const ChannelClose) callconv(.c) c_int {
    const c = maybe_c orelse return err_bad_state;
    const cl = maybe_cl orelse return err_bad_state;
    if (c.state != state_active and c.state != state_expired) return err_bad_state;
    if (!std.mem.eql(u8, c.channel_id[0..], cl.channel_id[0..])) return err_bad_id;
    if (cl.final_seq != c.current_seq or cl.final_device_share != c.device_share) return err_seq_match;
    c.state = state_closed;
    return ok;
}

pub export fn cm_channel_tick_expiry(maybe_c: ?*Channel, now_ms: u64) callconv(.c) void {
    const c = maybe_c orelse return;
    if (c.state == state_active and now_ms > c.expiry_ms) {
        c.state = state_expired;
    }
}

pub export fn cm_channel_validate_utxo_ref(maybe_channel_id: ?[*]const u8, maybe_txid_display: ?[*]const u8) callconv(.c) bool {
    const channel_id = maybe_channel_id orelse return false;
    const txid_display = maybe_txid_display orelse return false;
    return std.mem.eql(u8, channel_id[0..16], txid_display[0..16]);
}
